File: nxflix-agents-ts/src/models/progress.ts
import { z } from 'zod';

export const SM2Data = z.object({
  easeFactor: z.number().min(1.3).default(2.5),
  interval: z.number().int().min(0).default(0),
  repetitions: z.number().int().min(0).default(0),
  nextReviewDate: z.string().datetime().nullish(),
  lastReviewDate: z.string().datetime().nullish(),
});
export type SM2Data = z.infer<typeof SM2Data>;

export const UserProgress = z.object({
  userId: z.string(),
  grammarId: z.string(),
  sm2Data: SM2Data.default({}),
  timesStudied: z.number().int().default(0),
  timesCorrect: z.number().int().default(0),
  lastScore: z.number().nullish(),
  masteryLevel: z.number().int().min(0).max(5).default(0),
  createdAt: z.string().datetime().optional(),
  updatedAt: z.string().datetime().optional(),
});
export type UserProgress = z.infer<typeof UserProgress>;

export const SessionResult = z.object({
  grammarId: z.string(),
  questionsAsked: z.number().int(),
  correctAnswers: z.number().int(),
  score: z.number().min(0).max(5),
});
export type SessionResult = z.infer<typeof SessionResult>;

export const StudySession = z.object({
  id: z.string(),
  userId: z.string(),
  grammarIds: z.array(z.string()),
  startedAt: z.string().datetime().optional(),
  completedAt: z.string().datetime().nullish(),
  results: z.array(SessionResult).default([]),
  totalQuestions: z.number().int().default(0),
  totalCorrect: z.number().int().default(0),
});
export type StudySession = z.infer<typeof StudySession>;

File: nxflix-agents-python/src/nxflix_agents/models/progress.py
"""User progress and study session models."""

from datetime import datetime
from pydantic import BaseModel, Field


class SM2Data(BaseModel):
    """SM-2 spaced repetition algorithm data."""

    ease_factor: float = Field(default=2.5, ge=1.3)
    interval: int = Field(default=0, ge=0)  # Days until next review
    repetitions: int = Field(default=0, ge=0)
    next_review_date: datetime | None = None
    last_review_date: datetime | None = None


class UserProgress(BaseModel):
    """User's progress on a specific grammar point."""

    user_id: str
    grammar_id: str
    sm2_data: SM2Data = Field(default_factory=SM2Data)
    times_studied: int = 0
    times_correct: int = 0
    last_score: float | None = None
    mastery_level: int = Field(default=0, ge=0, le=5)  # 0-5 mastery scale
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    @property
    def accuracy(self) -> float:
        """Calculate accuracy percentage."""
        if self.times_studied == 0:
            return 0.0
        return (self.times_correct / self.times_studied) * 100


class SessionResult(BaseModel):
    """Result of a single grammar point within a study session."""

    grammar_id: str
    questions_asked: int
    correct_answers: int
    score: float  # 0-5 quality score for SM-2


class StudySession(BaseModel):
    """A study session tracking user's practice."""

    id: str
    user_id: str
    grammar_ids: list[str]
    started_at: datetime = Field(default_factory=datetime.utcnow)
    completed_at: datetime | None = None
    results: list[SessionResult] = []
    total_questions: int = 0
    total_correct: int = 0

    @property
    def is_complete(self) -> bool:
        """Check if session is complete."""
        return self.completed_at is not None

    @property
    def accuracy(self) -> float:
        """Calculate session accuracy."""
        if self.total_questions == 0:
            return 0.0
        return (self.total_correct / self.total_questions) * 100
